mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,416 @@
+use mmap::{mmap_read, mmap_write};
+
+pub type RingBufferResult<T> = Result<T, error::RingBufferError>;
+
+#[derive(Default)]
+pub(crate) struct FilePointers {
+    write: AtomicUsize,
+    read: AtomicUsize,
+}
+
+/// Imagine there are three pointers:
+/// W - A write pointer
+/// R - A read pointer
+/// D - A delete pointer
+///
+/// With a file buffer similar to below:
+/// +---+---+---+---+---+---+---+---+---+
+/// | A | B | C | D | E | F | G | H | I |
+/// +---+---+---+---+---+---+---+---+---+
+///
+/// It is possible that W could wrap around
+/// and cause us to lose data due to variable
+/// sized data being written.
+/// If W + data size > R and should write precede read,
+/// the data could become corrupted.
+/// Either this is prevented by scanning the file for
+/// the next block and checking if the write will fit.
+/// Or, W + data must not pass both D and R.
+/// Not passing D helps with guaranteeing all data read was handled.
+/// Not passing R helps avoid corruption.
+pub struct RingBuffer {
+    inner: Arc<Mutex<MmapMut>>,
+    max_file_size: usize,
+    flush_options: FlushOptions,
+    pointers: Arc<FilePointers>,
+    waker: Option<Waker>,
+    flush_state: Arc<FlushState>,
+}
+
+unsafe impl Send for RingBuffer {}
+unsafe impl Sync for RingBuffer {}
+
+fn create_file(file_name: String) -> IOResult<File> {
+    OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .open(&PathBuf::from(file_name))
+}
+
+impl RingBuffer {
+    pub fn new(file_name: String, max_file_size: usize, flush_options: FlushOptions) -> Self {
+        let file = create_file(file_name).expect("Failed to create file");
+        file.set_len(max_file_size as u64)
+            .expect("Failed to set file size");
+        let mmap = unsafe { MmapMut::map_mut(&file).expect("Failed to open mmap") };
+        Self {
+            inner: Arc::new(Mutex::new(mmap)),
+            max_file_size,
+            flush_options,
+            pointers: Arc::new(FilePointers::default()),
+            waker: None,
+            flush_state: Arc::new(FlushState::new()),
+        }
+    }
+}
+
+fn should_flush(flush_options: FlushOptions, flush_state: &FlushState) -> bool {
+    match flush_options {
+        FlushOptions::AfterEachWrite => true,
+        FlushOptions::AfterXWrites(xwrites) => flush_state.writes.load(Ordering::SeqCst) >= xwrites,
+        FlushOptions::AfterXBytes(xbytes) => {
+            flush_state.bytes_written.load(Ordering::SeqCst) >= xbytes
+        }
+        FlushOptions::AfterXMilliseconds(xmillis_elapsed) => {
+            flush_state.millis_elapsed.load(Ordering::SeqCst) >= xmillis_elapsed
+        }
+        FlushOptions::Off => false,
+    }
+}
+
+struct RingBufferWriteFuture {
+    pointers: Arc<FilePointers>,
+    flush_options: FlushOptions,
+    mmap: Arc<Mutex<MmapMut>>,
+    flush_state: Arc<FlushState>,
+    value: Vec<u8>,
+    max_file_size: usize,
+}
+
+impl Future for RingBufferWriteFuture {
```

Comment:
Each operation is implemented as a future, that way we can do two things:
1. apply back pressure to read or write if either one is moving faster. By this I mean, if we are writing a lot and not reading as much or vice versa, we will be in a pending state in the async execution and give the other operation a chance to catch up. Not sure how often pending is checked, though, so that needs to be thought about.
2. The other part about making these futures is it allows us to be prepared to switch to anything like AsyncWrite or other async traits with little effort.